Source/NSGridView.m

```diff
@@ -0,0 +1,1055 @@
+/* Implementation of class NSGridView
+   Copyright (C) 2020 Free Software Foundation, Inc.
+   
+   By: Gregory John Casamento
+   Date: 08-08-2020
+
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSArray.h>
+#import "AppKit/NSGridView.h"
+#import "GSFastEnumeration.h"
+
+@implementation NSGridView
+
++ (void) initialize
+{
+  if (self == [NSGridView class])
+    {
+      [self setVersion: 1];
+    }
+}
+
+- (void) _refreshCells
+{
+  NSUInteger i = 0;
+  NSUInteger num_col = [self numberOfColumns];
+  // NSUInteger num_rows = [self numberOfRows];
+  NSRect f = [self frame];
+  CGFloat current_x = 0.0, current_y = f.size.height;
+  NSMutableArray *colWidths = [NSMutableArray array];
+  NSMutableArray *rowHeights = [NSMutableArray array];
+
+  // precalculate w/h
+  FOR_IN(NSGridCell*, c, _cells)
+    {
+      NSView *v = [c contentView];
+      NSUInteger ri = 0, ci = 0;
+      NSRect rect = NSZeroRect;
+      
+      // Get row and column index...
+      ci = i % num_col;
+      ri = i / num_col;
+      
+      if (v != nil)
+        {
+          rect = [v frame];
+          // NOTE: I am not sure why this is needed, the row and column heights are coming in via the nib
```

Comment:
Float laues that small should be treated as zero. Most likely the programmer that implemented coding on Cocoa was not aware of this and instead of not encoding that value encoded it. The best we can do now is to add a small delta to our comparisons. Base is already doing this in a few places (See almostEqual in NSGeometry.m). You could copy the code from there, but if the simple "> 0.0" is sufficient maybe that is already good enough? I am a bit unsure here as this may depend on the specific CGFloat implementation.

Reply:
I think you're correct.  The values for these are not in the XIB file.  I'm guessing the XIB->NIB compiler is simply instantiating the objects and then using the non-keyed methods to save them and since they are not treated as zero by the non-keyed code it is simply dumping them out.   

I've attached the XIB file here: [MainMenu.xib.txt](https://github.com/gnustep/libs-gui/files/6014922/MainMenu.xib.txt)

AS AN ASIDE: I have toyed with the idea of doing a XIB->gorm "compiler".  The trick would be to create an NSWindow subclass that would allow you to add views and such in memory so that no GUI would appear on the display during compilation, but since we can read XIBs directly it's hardly a priority at this point.

Reply:
Another thought: It's confusing that they are not in the xib, since they are clearly something that the InterfaceBuilder in XCode is setting up...  please see this here: 

![column_editing](https://user-images.githubusercontent.com/27943/108595412-bcac0c00-734d-11eb-8598-14ecd751ea4a.png)

If you notice, the column editor has "100" and the "explicit" flag is unchecked.   Similarly if I click on a row I get the value of "30" for height.  I believe these MIGHT be defaults, but I am not sure.  I am going to operate on the assumption that they are and if the method I wrote for finding a "prototype view" to base the height/width on, I will use that.     

Also, at this point, I may leave the "merge..." method unimplemented as I am not sure I understand yet how that will work.  I believe it may be just as simple as adding the two views together and just letting the math skip over the merged view (either vertically or horizontally as the case may be), but I am not sure.